skf/markdown/code_examples/java/1-code_example--CSRF_Token_JSF--.md

```diff
@@ -0,0 +1,179 @@
+SecureRandom csprng = new SecureRandom();
+
+Then we generate a long value token containing a high entropy
+byte[] randomBytes  = new byte[128];
+
+csprng.nextBytes(randombytes);
+
+Then we base64 encode the string
+String csrftoken = Base64.getEncoder().encodeToString(randomBytes);
+
+Then we set the session attribute.
+
+origRequest.getSession(false);
+origRequest.getSession().setAttribute("CSRF", CSRftoken);
+
+The next step is to implement this random token in each form field as a hidden input parameter
+and send it to a function which checks if the submitted token is equal to the one set after succesful validation.
+
+The following .xhtml snippet shows the code used to place the antiCSRF token inside the page.
+When the page renders, the <cu:antiCSRF/> is created as a viewstate encoded html input tag
+which then carries the antiCSRF token.
+While in process of rendering the page, a new token is generated
+and added into the existing session.
+When the user press the commandButton
+then CSRF token parameter is compared with the CSRF session parameter. 
+
+*/
+
+
+/*
+          <f:view contentType="text/html">    
+          <f:event listener="#{userLoginView.isAuthenticated}" type="preRenderView" />
+[ .... ]
+
+          <p:commandButton action="password?faces-redirect=true" value="Add User" ajax="false">
+                <cu:antiCSRF/>
+    </p:commandButton>
+              </h:form>
+
+[ .... ]
+
+/* 
+
+the following function used to generate the new Session which then is added to the already existing session. 
+
+*/
+
+
+public void generateToken(){
+		
+		HttpServletRequest origRequest = (HttpServletRequest)FacesContext.getCurrentInstance().getExternalContext().getRequest();
+
+    	//we include the random password/token class.
+  	    Randomizer CSRF = new Randomizer();
+		 /*
+        Now we create a random value for our CSRF tokens. See "Random password token generation" in
+        the code examples for more detailed information:
+        */
+        String CSRftoken = CSRF.generate(25);
+                
+        //Set an accesor session.
+        origRequest.getSession(false);
+        origRequest.getSession().setAttribute("CSRF", CSRftoken);
+	}
+
+/* 
+the following function used to destroy the cookie and invalidate the session when the CSRF tokens dont match 
+*/
+
+    public void antiCSRF() throws IOException
+    {		
+        HttpServletRequest origRequest = (HttpServletRequest)FacesContext.getCurrentInstance().getExternalContext().getRequest();
+        HttpServletResponse origResponse = (HttpServletResponse)FacesContext.getCurrentInstance().getExternalContext().getResponse();
+        String AUTH_KEY =  (String) FacesContext.getCurrentInstance().getExternalContext().getSessionMap().get("AUTH_KEY");
+       	FacesContext.getCurrentInstance().getExternalContext().getSessionMap().remove(AUTH_KEY);
+   		FacesContext.getCurrentInstance().getExternalContext().invalidateSession();
```

Comment:
The code fomr line 90 to 94 could be replaced by this one:
```
ExternalContext externalContext = FacesContext.getCurrentInstance().getExternalContext();

HttpServletRequest origRequest = (HttpServletRequest)externalContext.getRequest();

HttpServletResponse origResponse = (HttpServletResponse)externalContext.getResponse();

String AUTH_KEY =  (String) externalContext.getSessionMap().get("AUTH_KEY");

externalContext.getSessionMap().remove(AUTH_KEY);

externalContext.invalidateSession();
```
to keep the code clear.